File: ghostcam-py/ghostcam/firmware.py
"""Firmware self-update.

Mirrors camera/firmware.go. Polls /api/v1/firmware/latest, verifies the
sha256 of the downloaded artifact, stages it as `staged-update.deb` in
the data dir, and signals the caller (which exits) so systemd's
ExecStartPre can install it.

`VERSION` is set at build time via the wheel/deb. The sentinel "dev"
disables the firmware check entirely — same convention as the Go
camera's `var Version = "dev"`.
"""

from __future__ import annotations

import hashlib
import logging
from pathlib import Path

import httpx
from pydantic import BaseModel, Field

from ghostcam.client import VERSION, Client

logger = logging.getLogger(__name__)


class FirmwareRelease(BaseModel):
    version: str
    download_url: str
    sha256: str = ""


class FirmwareResponse(BaseModel):
    release: FirmwareRelease | None = Field(default=None)


async def check_firmware_update(client: Client, data_dir: Path) -> bool:
    """Returns True if a new firmware was staged. The caller should exit
    so systemd restarts the service."""
    if VERSION == "dev":
        logger.debug("firmware check skipped (dev build)")
        return False

    logger.info("checking for firmware update (current=%s)", VERSION)

    try:
        resp = await _get_firmware_latest(client)
    except Exception as e:  # noqa: BLE001
        logger.warning("firmware check failed: %s", e)
        return False

    if resp.release is None:
        logger.debug("no firmware release available")
        return False

    if resp.release.version == VERSION:
        logger.debug("firmware up to date (%s)", VERSION)
        return False

    logger.info(
        "new firmware available: %s -> %s",
        VERSION, resp.release.version,
    )

    staged = data_dir / "staged-update.deb"
    try:
        await _download_to_file(resp.release.download_url, staged)
    except Exception as e:  # noqa: BLE001
        logger.error("firmware download failed: %s", e)
        return False

    if resp.release.sha256:
        actual = _sha256_file(staged)
        if actual != resp.release.sha256:
            logger.error("firmware hash mismatch (expected=%s actual=%s)",
                         resp.release.sha256, actual)
            try:
                staged.unlink()
            except OSError:
                pass
            return False
        logger.info("firmware hash verified: %s", actual)

    logger.info("firmware staged, restart will apply: %s", staged)
    return True


async def _get_firmware_latest(client: Client) -> FirmwareResponse:
    url = client.server_url + "/api/v1/firmware/latest"
    http = await client._client()
    resp = await http.get(url, timeout=10.0)
    if resp.status_code != 200:
        raise RuntimeError(f"firmware/latest returned {resp.status_code}")
    return FirmwareResponse.model_validate(resp.json())


async def _download_to_file(url: str, dest: Path) -> None:
    tmp = dest.with_suffix(dest.suffix + ".tmp")
    async with httpx.AsyncClient(timeout=300.0) as http:
        async with http.stream("GET", url) as resp:
            if resp.status_code != 200:
                raise RuntimeError(f"download returned {resp.status_code}")
            with tmp.open("wb") as f:
                async for chunk in resp.aiter_bytes(65536):
                    f.write(chunk)
    tmp.replace(dest)


def _sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as f:
        while chunk := f.read(65536):
            h.update(chunk)
    return h.hexdigest()

File: camera/internal/firmware/firmware.go
package firmware

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"os"
	"path/filepath"
	"time"
)

// Client is the firmware updater's narrow view of the server HTTP
// client. The concrete *main.Client (camera/client.go) satisfies this
// surface; defining the interface here lets firmware live in its own
// subpackage without importing package main.
type Client interface {
	// HTTPClient returns the underlying *http.Client; firmware uses it
	// directly for the (currently unauthenticated) GET /firmware/latest
	// call and the .deb download.
	HTTPClient() *http.Client
	// ServerURL returns the trimmed server base URL (no trailing slash).
	ServerURL() string
}

type firmwareRelease struct {
	Version     string `json:"version"`
	DownloadURL string `json:"download_url"`
	SHA256      string `json:"sha256"`
}

type firmwareResponse struct {
	Release *firmwareRelease `json:"release"`
}

// CheckFirmwareUpdate checks the server for a newer firmware version.
// If found, downloads the .deb to {dataDir}/staged-update.deb and returns
// true. The caller should exit so systemd restarts. ExecStartPre backs up
// the current binary and runs dpkg -i; if the new binary fails to write
// boot_ok (crash-loop), the next restart rolls back to .prev.
//
// Invoked only via the telemetry-driven `update_firmware` command path
// (camera/commands.go) — never at startup. The eager startup call was
// removed in f87a0bf because boot_ok is written after first telemetry
// success, so an eager exit at boot raced the rollback gate.
//
// version is the running binary's main.Version. Passed as a parameter
// (rather than read from package main directly) because internal
// subpackages can't import main, and the -X main.Version=$SHA
// reproducible-build gate requires Version to stay there.
func CheckFirmwareUpdate(ctx context.Context, client Client, dataDir, version string) bool {
	if version == "dev" {
		slog.Debug("firmware check skipped (dev build)")
		return false
	}

	slog.Info("checking for firmware update", "current_version", version)

	resp, err := getFirmwareLatest(ctx, client)
	if err != nil {
		slog.Warn("firmware check failed", "error", err)
		return false
	}

	if resp.Release == nil {
		slog.Debug("no firmware release available")
		return false
	}

	if resp.Release.Version == version {
		slog.Debug("firmware is up to date", "version", version)
		return false
	}

	slog.Info("new firmware available", "current", version, "new", resp.Release.Version)

	stagedPath := filepath.Join(dataDir, "staged-update.deb")
	if err := downloadToFile(ctx, resp.Release.DownloadURL, stagedPath); err != nil {
		slog.Error("firmware download failed", "error", err)
		return false
	}

	// Verify SHA256 if server provided a hash (backward-compat: skip if empty)
	if resp.Release.SHA256 != "" {
		actual, err := fileHash(stagedPath)
		if err != nil {
			slog.Error("firmware hash computation failed", "error", err)
			os.Remove(stagedPath)
			return false
		}
		if actual != resp.Release.SHA256 {
			slog.Error("firmware hash mismatch, discarding",
				"expected", resp.Release.SHA256, "actual", actual)
			os.Remove(stagedPath)
			return false
		}
		slog.Info("firmware hash verified", "sha256", actual)
	}

	slog.Info("firmware staged, restarting for install", "new_version", resp.Release.Version, "staged", stagedPath)
	return true
}

func getFirmwareLatest(ctx context.Context, client Client) (*firmwareResponse, error) {
	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()

	url := fmt.Sprintf("%s/api/v1/firmware/latest", client.ServerURL())
	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, err
	}

	resp, err := client.HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("firmware check returned %d", resp.StatusCode)
	}

	var result firmwareResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("decoding firmware response: %w", err)
	}
	return &result, nil
}

func fileHash(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()
	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}
	return hex.EncodeToString(h.Sum(nil)), nil
}

func downloadToFile(ctx context.Context, url, destPath string) error {
	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return err
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("downloading firmware: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("download returned %d", resp.StatusCode)
	}

	tmpPath := destPath + ".tmp"
	f, err := os.OpenFile(tmpPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
	if err != nil {
		return fmt.Errorf("creating temp file: %w", err)
	}

	n, err := io.Copy(f, resp.Body)
	f.Close()
	if err != nil {
		os.Remove(tmpPath)
		return fmt.Errorf("writing firmware: %w", err)
	}

	if err := os.Rename(tmpPath, destPath); err != nil {
		os.Remove(tmpPath)
		return fmt.Errorf("staging firmware: %w", err)
	}

	slog.Info("firmware downloaded", "size_bytes", n, "path", destPath)
	return nil
}
